File: cryptcow.cpp
/*
ID: wrwwctb1
TASK: cryptcow
LANG: C++14
*/

#include <iostream>
#include <string>
#include <fstream>
#include <sstream>
#include <cassert>

#include <vector>
#include <deque>
#include <list>
#include <set> // has multiset
#include <map> // has multimap
#include <unordered_set> // has unordered_multiset
#include <unordered_map> // has unordered_multimap
#include <stack>
#include <queue> // has priority_queue

#include <algorithm>
#include <cmath>
#include <cfloat>
#include <ctime>
#include <climits>
#include <iomanip>
#include <cstring>

#define FOR( i, N ) for( int (i) = 0; (i) < (N); (i)++ )
//#define FOR_( i, start, endplusone ) for( int (i) = (start); (i) < (endplusone); (i)++ )
const double EPS = 1e-6;
const double PI = 3.141592653589793;
using namespace std;
// profile: 27196474
















const int HASH_NUM = 10000000;
bool hashVisited[HASH_NUM];

unsigned int elf_hash(string str){
    unsigned long h = 0, g, i, l;
    l = str.length();
    for (i = 0; i < l; i ++){
        h = (h << 4) + str[i];
        if (g = h & 0xf0000000l)
            h ^= g >> 24;
        h &= ~g;
    }
    return h % HASH_NUM;

    // // slowest
    // hash<int> hasher;
    // size_t seed = 0;
    // FOR (i, str.size())
    //     seed ^= hasher(str[i]) + 0x9e3779b9 + (seed<<6) + (seed>>2);
    // return seed % HASH_NUM;

    // // slower
    // unsigned long hash = 5381;
    // FOR (i, str.size())
    //     hash = (hash << 5) + hash + str[i]; /* hash * 33 + c */
    // return hash % HASH_NUM;
}

void str2map(string given, unordered_map<char, int> &cntGiven){
    for (char ch: given)
        cntGiven[ch] += 1;
}

bool checkLetterCounts(string target, string given){
    if (given.size() < target.size())
        return false;
    int delta = given.size() - target.size();
    if (delta % 3)
        return false;
    unordered_map<char, int> cntTarget;
    unordered_map<char, int> cntGiven;
    str2map(target, cntTarget);
    str2map(given, cntGiven);

    if (delta){
        cntTarget['C'] = delta / 3;
        cntTarget['O'] = delta / 3;
        cntTarget['W'] = delta / 3;
    }
    if (cntTarget != cntGiven)
        return false;
    return true;
}

bool recurse(string &target, string &given){


    int tl = target.size();
    int gl = given.size();
    if (tl == gl)
        return target == given;

    int C0 = given.find('C');
    int W1 = given.rfind('W');
    // assert (C0 >= 0 && W1 >= 0);

    int targetSubEnd = tl-(gl-W1)+1;
    if (targetSubEnd < 0 || tl < targetSubEnd)
        return false;
    if (target.substr(0, C0) != given.substr(0, C0))
        return false;
    if (target.substr(targetSubEnd, tl) != given.substr(W1+1))
        return false;

    string trimmedTarget = target.substr(C0, targetSubEnd-C0);


    int Cs[9];
    int Os[9];
    int Ws[9];
    int Ccnt = 0;
    int Ocnt = 0;
    int Wcnt = 0;
    int prev = -1;
    FOR (i, gl){
        char ch = given[i];
        if (ch == 'C' || ch == 'O' || ch == 'W'){
            if (ch == 'C')
                Cs[Ccnt++] = i;
            else if (ch == 'O')
                Os[Ocnt++] = i;
            else
                Ws[Wcnt++] = i;

            if (prev > -1 && trimmedTarget.find(given.substr(prev+1, i-prev-1)) == string::npos)
                return false;
            prev = i;
        }
    }

    FOR (io, Ocnt){
        FOR (ic, Ccnt){
            FOR (iw, Wcnt){
                int C = Cs[ic];
                int O = Os[io];
                int W = Ws[iw];
                if (C < O && O < W){


                    string cand = given.substr(C0, C-C0) +
                                  given.substr(O+1, W-O-1) +
                                  given.substr(C+1, O-C-1) +
                                  given.substr(W+1, W1-W);

                    int hash = elf_hash(cand);
                    if ( not hashVisited[hash] ){
                        hashVisited[hash] = true;

                        if (recurse(trimmedTarget, cand))
                            return true;
                    }
                }
            }
        }
    }
    return false;
}

void output(ofstream &fout, int a, int b){
    fout << a << ' ' << b << endl;
    cout << '.' << a << ' ' << b << endl;
}

int main(int argc, char **argv){
    time_t ttt = time(nullptr);


    const string filename("cryptcow");


    string filenameo = filename + ".out";
    string filenamei = filename + ".in";
    ofstream fout(filenameo.c_str());
    ifstream fin(filenamei.c_str());



    string target = "Begin the Escape execution at the Break of Dawn";
    string given;
    getline(fin, given);





    if (!checkLetterCounts(target, given))
        output(fout, 0, 0);
    else{
        if (recurse(target, given))
            output(fout, 1, (given.size() - target.size()) / 3);
        else
            output(fout, 0, 0);
    }












    printf("~%d\n", time(nullptr) - ttt);
    //system("pause");
    return 0;
}

File: cryptcow.py
"""
ID: wrwwctb1
LANG: PYTHON3
TASK: cryptcow
6.3.2

review
search with pruning
most critical:
any substring between any of {C, O, W} must be an original unaltered substring

interesting speed trade off

test data:
BeCOgC CC execuOf DOBCiCCrWaOOt theCOCeak oWWin Oon aWtheOOW EscapeWtWWWwn


file       main feature                              performance
.py        set()                                      6s
.py1       elf hash wo collision handling            17s
.cpp1      char *, unordered_set                      6s
.cpp       string, elf hash wo collision handling     2s
"""

import os, sys, re, time
from collections import Counter, deque, defaultdict
from queue import Queue
from copy import copy, deepcopy
from itertools import combinations, permutations, accumulate, \
                      combinations_with_replacement
from functools import lru_cache, cmp_to_key, partial as functools_partial
from heapq import heappush, heappop, nlargest, nsmallest
from bisect import bisect_left, bisect_right
from math import ceil, floor, factorial, gcd, modf, log, log2, log10, sqrt, \
         pi, sin, cos, tan, asin, acos, atan, atan2, hypot, erf, erfc, inf, nan
# sys.setrecursionlimit(5782)
# python -m cProfile -s time ha.py

print = functools_partial(print, flush=True)
input = lambda: fin.readline().strip('\n')

def printwrite(string):
    print(string)
    fout.write(str(string) + '\n')

filename = 'cryptcow'
fin = open(filename + '.in', 'r')
fout = open(filename + '.out', 'w')











def checkLetterCounts(target, given):
    lenTarget = len(target)
    lenGiven = len(given)
    if lenGiven < lenTarget:
        return False
    delta = lenGiven - lenTarget
    mult, rem = divmod(delta, 3)
    if rem != 0:
        return False
    cntTarget = Counter(target)
    cntGiven = Counter(given)

    if delta > 0:
        cntTarget['C'] = mult
        cntTarget['O'] = mult
        cntTarget['W'] = mult

    if cntTarget != cntGiven:
        return False
    return True

def recurse(target, given, subs, seen):
    tl = len(target)
    gl = len(given)
    if tl == gl:
        return target == given

    # compare before first C and after last W
    # this covers cases where first ind is not C or last ind is not W
    C0 = given.index('C')
    W1 = given.rindex('W')
    targetSubStart = tl-(gl-W1)+1
    if targetSubStart < 0 or tl < targetSubStart:
        return False
    if target[:C0] != given[:C0]:
        return False
    if target[targetSubStart:] != given[W1+1:]:
        return False

    trimmedTarget = target[C0:tl-(gl-W1)+1]

    # find all C O W
    # check all sub strings between each C O W
    Cs = []
    Os = []
    Ws = []
    prev = -1
    for i, ch in enumerate(given):
        if ch == 'C' or ch == 'O' or ch == 'W':
            if ch == 'C':
                Cs.append(i)
            elif ch == 'O':
                Os.append(i)
            else:
                Ws.append(i)

            # substrings between any C W O need to be in target
            if given[prev+1:i] not in subs:
                return False
            prev = i

    # generate possible rewinds, recurse
    for O in Os:
        for C in Cs:
            for W in Ws:
                if C < O and O < W:
                    cand = (given[C0:C]  +
                            given[O+1:W] +
                            given[C+1:O] +
                            given[W+1:W1+1])
                    if cand not in seen:
                        seen.add(cand)
                        if recurse(trimmedTarget, cand, subs, seen):
                            return True
    return False






tt = time.time()

target = 'Begin the Escape execution at the Break of Dawn'
given = input()


if not checkLetterCounts(target, given):
    printwrite('%d %d' % (0, 0))
else:
    # find set of substrings
    subs = set([''])
    for i in range(len(target)):
        for j in range(i+1, len(target)+1):
            subs.add(target[i:j])

    seen = set()
    if recurse(target, given, subs, seen):
        printwrite('%d %d' % (1, (len(given)-len(target))//3))
    else:
        printwrite('%d %d' % (0, 0))

print(time.time()-tt)









fin.close()
fout.close()
del print, input

